src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -656,80 +666,84 @@ object Gen extends GenArities with GenVersionSpecific {
-  def listOfN[T](n: Int, g: Gen[T]) = buildableOfN[List[T],T](n,g)
+  def listOfN[T](n: Int, g: Gen[T]) = buildableOfN[List[T], T](n, g)
 
   /** Generates a map of random length. The maximum length depends on the
    *  size parameter. This method is equal to calling
    *  <code>containerOf[Map,(T,U)](g)</code>. */
-  def mapOf[T,U](g: => Gen[(T,U)]) = buildableOf[Map[T,U],(T,U)](g)
+  def mapOf[T, U](g: => Gen[(T, U)]) = buildableOf[Map[T, U], (T, U)](g)
 
   /** Generates a non-empty map of random length. The maximum length depends
    *  on the size parameter. This method is equal to calling
    *  <code>nonEmptyContainerOf[Map,(T,U)](g)</code>. */
-  def nonEmptyMap[T,U](g: => Gen[(T,U)]) = nonEmptyBuildableOf[Map[T,U],(T,U)](g)
+  def nonEmptyMap[T,U](g: => Gen[(T,U)]) = nonEmptyBuildableOf[Map[T, U],(T, U)](g)
 
   /** Generates a map with at most the given number of elements. This method
    *  is equal to calling <code>containerOfN[Map,(T,U)](n,g)</code>. */
-  def mapOfN[T,U](n: Int, g: Gen[(T,U)]) = buildableOfN[Map[T,U],(T,U)](n,g)
+  def mapOfN[T,U](n: Int, g: Gen[(T, U)]) = buildableOfN[Map[T, U],(T, U)](n, g)
 
   /** Generates an infinite stream. */
   def infiniteStream[T](g: => Gen[T]): Gen[Stream[T]] = {
-    def unfold[A, S](z: S)(f: S => Option[(A, S)]): Stream[A] = f(z) match {
-      case Some((h, s)) => h #:: unfold(s)(f)
-      case None => Stream.empty
+    def unfold(p: P, seed: Seed): Stream[T] = {
+      val r = g.doPureApply(p, seed)
+      r.retrieve match {
+        case Some(t) => t #:: unfold(p, r.seed)
+        case None => Stream.empty
+      }
     }
     gen { (p, seed0) =>
       new R[Stream[T]] {
-        val result: Option[Stream[T]] = Some(unfold(seed0)(s => Some(g.pureApply(p, s) -> s.next)))
-        val seed: Seed = seed0.next
+        val result: Option[Stream[T]] = Some(unfold(p, seed0))
+        val seed: Seed = seed0.slide
       }
     }
   }
 
   /** A generator that picks a random number of elements from a list */
-  def someOf[T](l: Iterable[T]) = choose(0,l.size).flatMap(pick(_,l))
+  def someOf[T](l: Iterable[T]): Gen[collection.Seq[T]] =
+    choose(0, l.size).flatMap(pick(_,l))
 
   /** A generator that picks a random number of elements from a list */
-  def someOf[T](g1: Gen[T], g2: Gen[T], gs: Gen[T]*) =
+  def someOf[T](g1: Gen[T], g2: Gen[T], gs: Gen[T]*): Gen[collection.Seq[T]] =
     choose(0, gs.length+2).flatMap(pick(_, g1, g2, gs: _*))
 
   /** A generator that picks at least one element from a list */
-  def atLeastOne[T](l: Iterable[T]) = {
+  def atLeastOne[T](l: Iterable[T]): Gen[collection.Seq[T]] = {
     require(l.size > 0, "There has to be at least one option to choose from")
     choose(1,l.size).flatMap(pick(_,l))
   }
 
   /** A generator that picks at least one element from a list */
-  def atLeastOne[T](g1: Gen[T], g2: Gen[T], gs: Gen[T]*) =
+  def atLeastOne[T](g1: Gen[T], g2: Gen[T], gs: Gen[T]*): Gen[collection.Seq[T]] =
```

Comment:
why did you use `collection.Seq` here, but `Seq` below on `pick`?

Reply:
The existing code was already using `collection.Seq` -- I just made the types explicit. It's not possible to change these to `Seq` without changing the relevant `pick` definition to return `Seq`, breaking binary compatibility.

Reply:
Actually you can. Types are erased so the type parameter of Gen is not in the java binary type.

But in any case, this method calls pick and pick returns Seq, so how can this be a different type than pick?